--- a/src/args.c
+++ b/src/args.c
@@ -30,6 +30,8 @@
 #include "dhcp-discover.h"
 // defined in dnsmasq.c
 extern void print_dnsmasq_version(void);
+extern void do_version(void);
+extern void do_usage(void);
 
 // defined in database/shell.c
 extern int sqlite3_shell_main(int argc, char **argv);
@@ -141,19 +143,6 @@ void parse_args(int argc, char* argv[])
 			// Remember that the rest is for dnsmasq ...
 			consume_for_dnsmasq = true;
 
-			// Special command interpretation for "pihole-FTL -- --help dhcp"
-			if(argc > 1 && strcmp(argv[argc-2], "--help") == 0 && strcmp(argv[argc-1], "dhcp") == 0)
-			{
-				display_opts();
-				exit(EXIT_SUCCESS);
-			}
-			// and "pihole-FTL -- --help dhcp6"
-			if(argc > 1 && strcmp(argv[argc-2], "--help") == 0 && strcmp(argv[argc-1], "dhcp6") == 0)
-			{
-				display_opts6();
-				exit(EXIT_SUCCESS);
-			}
-
 			// ... and skip the current argument ("--")
 			continue;
 		}
@@ -162,6 +151,25 @@ void parse_args(int argc, char* argv[])
 		// dnsmasq
 		if(consume_for_dnsmasq)
 		{
+			if(strcmp(argv[i], "-v") == 0 || strcmp(argv[i], "--version") == 0)
+			{
+				do_version();
+				exit(EXIT_SUCCESS);
+			}
+			else if(strcmp(argv[i], "-w") == 0 || strcmp(argv[i], "--help") == 0)
+			{
+				if(strcmp(argv[argc-1], "dhcp") == 0)
+					display_opts();
+#ifdef HAVE_DHCP6
+				else if(strcmp(argv[argc-1], "dhcp6") == 0)
+					display_opts6();
+#endif
+				else
+					do_usage();
+
+				exit(EXIT_SUCCESS);
+			}
+
 			if(argv_dnsmasq != NULL)
 				free(argv_dnsmasq);
 
--- a/src/dnsmasq/dnsmasq.c
+++ b/src/dnsmasq/dnsmasq.c
@@ -2228,6 +2228,15 @@ void print_dnsmasq_version(void)
 {
   printf("****************************** dnsmasq ******************************\n");
   printf(_("Version:         %s\n"), VERSION);
-  printf(_("Compile options: %s\n\n"), compile_opts);
+  printf(_("Compile time options: %s\n\n"), compile_opts);
+}
+
+void do_version(void)
+{
+  printf(_("Dnsmasq version %s  %s\n"), VERSION, COPYRIGHT);
+  printf(_("Compile time options: %s\n\n"), compile_opts);
+  printf("This software comes with ABSOLUTELY NO WARRANTY.\n");
+  printf("Dnsmasq is free software, and you are welcome to redistribute it\n");
+  printf("under the terms of the GNU General Public License, version 2 or 3.\n");
 }
 /**************************************************************************************/
--- a/src/dnsmasq/option.c
+++ b/src/dnsmasq/option.c
@@ -774,7 +774,7 @@ static void add_txt(char *name, char *tx
 }
 #endif
 
-static void do_usage(void)
+void do_usage(void)
 {
   char buff[100];
   int i, j;
