diff --git a/src/database/sqlite3.c b/src/database/sqlite3.c
index f717033d..df4cd751 100644
--- a/src/database/sqlite3.c
+++ b/src/database/sqlite3.c
@@ -210267,7 +210267,7 @@
     goto update_fail;
   }
   pBt = pTab->db->aDb[iDb].pBt;
-  if( pgno<1 || pBt==0 || pgno>(int)sqlite3BtreeLastPage(pBt) ){
+  if( pgno<1 || pBt==0 || pgno>(unsigned int)sqlite3BtreeLastPage(pBt) ){
     zErr = "bad page number";
     goto update_fail;
   }
