--- a/src/main.c
+++ b/src/main.c
@@ -149,12 +149,12 @@ int main (int argc, char *argv[])
 	// be terminating immediately
 	sleepms(250);
 
-	// Save new queries to database (if database is used)
-	if(config.database.maxDBdays.v.ui > 0)
-	{
-		export_queries_to_disk(true);
-		log_info("Finished final database update");
-	}
+	// Always save database to disk despite DBExport is true and not in /tmp
+	if(exit_code != RESTART_FTL_CODE && strncmp(config.files.database.v.s, "/tmp", 4))
+	{
+		if(export_queries_to_disk(true))
+			log_info("Finished writing database to disk");
+	}
 
 	cleanup(exit_code);
 
