--- a/src/database/query-table.c
+++ b/src/database/query-table.c
@@ -29,7 +29,6 @@
 #include "gc.h"
 
 static sqlite3 *_memdb = NULL;
-static bool store_in_database = false;
 static double new_last_timestamp = 0;
 static unsigned int new_total = 0, new_blocked = 0;
 static unsigned long last_mem_db_idx = 0, last_disk_db_idx = 0;
@@ -1371,11 +1370,6 @@ bool queries_to_database(void)
 		log_debug(DEBUG_DATABASE, "Not storing query in database as there are none");
 		return true;
 	}
-	if(!store_in_database)
-	{
-		log_debug(DEBUG_DATABASE, "Not storing query in database as this is disabled");
-		return true;
-	}
 
 	// Loop over recent queries and store new or changed ones in the
 	// in-memory database
@@ -1651,6 +1645,4 @@ static void load_queries_from_disk(void)
 
 	// Log some information about the imported queries (if any)
 	log_counter_info();
-
-	store_in_database = true;
 }
