--- a/advanced/Scripts/COL_TABLE
+++ b/advanced/Scripts/COL_TABLE
@@ -1,5 +1,12 @@
+# Check if tput exists
+if command -v tput &> /dev/null; then
+  tputCommand="tput colors"
+else
+  tputCommand="echo 8"
+fi
+
 # Determine if terminal is capable of showing colors
-if ([ -t 1 ] && [ $(tput colors) -ge 8 ]) || [ "${WEBCALL}" ]; then
+if [[ -t 1 ]] && [[ $(${tputCommand}) -ge 8 ]]; then
   # Bold and underline may not show up on all clients
   # If something MUST be emphasized, use both
   COL_BOLD='[1m'
