--- a/api.php
+++ b/api.php
@@ -33,10 +33,6 @@ if (isset($_GET['enable']) && $auth) {
 		send_signal_FTL(35);
 	}
 	$data = array_merge($data, array("status" => "enabled"));
-	if (file_exists("/var/lib/pihole/custom_disable_timer"))
-	{
-		unlink("/var/lib/pihole/custom_disable_timer");
-	}
 }
 elseif (isset($_GET['disable']) && $auth)
 {
@@ -50,26 +46,10 @@ elseif (isset($_GET['disable']) && $auth
 		// Skip token validation if explicit auth string is given
 		check_csrf($_GET['token']);
 	}
-	$disable = intval($_GET['disable']);
-	// intval returns the integer value on success, or 0 on failure
-	if($disable > 0)
-	{
-		$timestamp = time();
-		pihole_execute("disable ".$disable."s");
-		file_put_contents("/var/lib/pihole/custom_disable_timer",($timestamp+$disable)*1000);
-	}
-	else
-	{
-		uci_set("pihole", "@pihole[0]", "blocking_enabled", 0);
-		if(uci_commit("pihole")) {
-			send_signal_FTL(35);
-		}
-		if (file_exists("/var/lib/pihole/custom_disable_timer"))
-		{
-			unlink("/var/lib/pihole/custom_disable_timer");
-		}
+	uci_set("pihole", "@pihole[0]", "blocking_enabled", 0);
+	if(uci_commit("pihole")) {
+		send_signal_FTL(35);
 	}
-
 	$data = array_merge($data, array("status" => "disabled"));
 }
 elseif (isset($_GET['versions']))
--- a/scripts/pi-hole/js/footer.js
+++ b/scripts/pi-hole/js/footer.js
@@ -65,7 +65,7 @@ function countDown() {
   }
 }
 
-function piholeChange(action, duration) {
+function piholeChange(action) {
   var token = encodeURIComponent($("#token").text());
   var enaT = $("#enableTimer");
   var btnStatus;
@@ -85,14 +85,10 @@ function piholeChange(action, duration)
     case "disable":
       btnStatus = $("#flip-status-disable");
       btnStatus.html("<i class='fa fa-spinner'> </i>");
-      $.getJSON("api.php?disable=" + duration + "&token=" + token, function (data) {
+      $.getJSON("api.php?disable&token=" + token, function (data) {
         if (data.status === "disabled") {
           btnStatus.html("");
           piholeChanged("disabled");
-          if (duration > 0) {
-            enaT.html(Date.now() + duration * 1000);
-            setTimeout(countDown, 100);
-          }
         }
       });
       break;
@@ -230,29 +226,11 @@ $(function () {
 $("#pihole-enable").on("click", function (e) {
   e.preventDefault();
   localStorage.removeItem("countDownTarget");
-  piholeChange("enable", "");
+  piholeChange("enable");
 });
-$("#pihole-disable-indefinitely").on("click", function (e) {
+$("#pihole-disable").on("click", function (e) {
   e.preventDefault();
-  piholeChange("disable", "0");
-});
-$("#pihole-disable-10s").on("click", function (e) {
-  e.preventDefault();
-  piholeChange("disable", "10");
-});
-$("#pihole-disable-30s").on("click", function (e) {
-  e.preventDefault();
-  piholeChange("disable", "30");
-});
-$("#pihole-disable-5m").on("click", function (e) {
-  e.preventDefault();
-  piholeChange("disable", "300");
-});
-$("#pihole-disable-custom").on("click", function (e) {
-  e.preventDefault();
-  var custVal = $("#customTimeout").val();
-  custVal = $("#btnMins").hasClass("active") ? custVal * 60 : custVal;
-  piholeChange("disable", custVal);
+  piholeChange("disable");
 });
 
 // Session timer
--- a/scripts/pi-hole/php/header.php
+++ b/scripts/pi-hole/php/header.php
@@ -236,7 +236,6 @@ if($auth) {
 ?>
 
 <!-- Send token to JS -->
-<div id="enableTimer" hidden><?php if(file_exists("/var/lib/pihole/custom_disable_timer")){ echo file_get_contents("/var/lib/pihole/custom_disable_timer"); } ?></div>
 <div class="wrapper">
     <header class="main-header">
         <!-- Logo -->
@@ -475,40 +474,9 @@ if($auth) {
                 </li>
                 <!-- Toggle -->
                 <li id="pihole-disable" class="treeview"<?php if ($pistatus == "0") { ?> hidden<?php } ?>>
-                  <a href="#">
-                    <i class="fa fa-fw menu-icon fa-stop"></i> <span>Disable&nbsp;&nbsp;&nbsp;<span id="flip-status-disable"></span></span>
-                    <span class="pull-right-container">
-                      <i class="fa fa-angle-left pull-right"></i>
-                    </span>
-                  </a>
-                  <ul class="treeview-menu">
-                    <li>
-                        <a href="#" id="pihole-disable-indefinitely">
-                            <i class="fa fa-fw menu-icon fa-infinity"></i> Indefinitely
-                        </a>
-                    </li>
-                    <li>
-                        <a href="#" id="pihole-disable-10s">
-                            <i class="fa fa-fw menu-icon fa-clock"></i> For 10 seconds
-                        </a>
-                    </li>
-                    <li>
-                        <a href="#" id="pihole-disable-30s">
-                            <i class="fa fa-fw menu-icon fa-clock"></i> For 30 seconds
-                        </a>
-                    </li>
-                    <li>
-                        <a href="#" id="pihole-disable-5m">
-                            <i class="fa fa-fw menu-icon fas fa-clock"></i> For 5 minutes
-                        </a>
-                    </li>
-                    <li>
-                      <a href="#" id="pihole-disable-cst" data-toggle="modal" data-target="#customDisableModal">
-                            <i class="fa fa-fw menu-icon fa-user-clock"></i> Custom time
-                      </a>
-                    </li>
-                  </ul>
-                    <!-- <a href="#" id="flip-status"><i class="fa fa-stop"></i> <span>Disable</span></a> -->
+                    <a href="#">
+                        <i class="fa fa-fw menu-icon fa-stop"></i> Disable<span id="flip-status-disable">
+                    </a>
                 </li>
                 <li id="pihole-enable" class="treeview"<?php if (!in_array($pistatus,["0","-1","-2"])) { ?> hidden<?php } ?>>
                     <a href="#">
