--- a/settings-dhcp.lp
+++ b/settings-dhcp.lp
@@ -15,6 +15,19 @@ mg.include('scripts/pi-hole/lua/settings
 ?>
 <div class="row">
     <!-- DHCP Settings Box -->
+    <div class="col-md-12 settings-level-basic">
+        <div class="box box-warning">
+            <div class="box-header with-border" style="text-align:center;">
+                <h3 class="box-title">INFORMATION&nbsp;<i class="fas fa-exclamation-triangle"></i></h3>
+            </div>
+            <div class="box-body">
+                <p style="text-align:center;">
+                    All DHCP Server Options has been disabled,
+                    please do configuration via Luci GUI or uci CLI.
+                </p>
+            </div>
+        </div>
+    </div>
     <div class="col-md-6 settings-level-basic">
         <div class="box box-warning">
             <div class="box-header with-border">
@@ -24,7 +37,7 @@ mg.include('scripts/pi-hole/lua/settings
                 <div class="row">
                     <div class="col-md-12">
                         <div>
-                            <input type="checkbox" id="dhcp.active" data-key="dhcp.active"><label for="dhcp.active"><strong>DHCP server enabled</strong></label>
+                            <input type="checkbox" id="dhcp.active" data-key="dhcp.active" disabled><label for="dhcp.active"><strong>DHCP server enabled</strong></label>
                             <p class="help-block" id="dhcpnotice" lookatme-text="Make sure your router's DHCP server is disabled when using the Pi-hole DHCP server!">Make sure your router's DHCP server is disabled when using the Pi-hole DHCP server!</p>
                         </div>
                     </div>
@@ -35,9 +48,9 @@ mg.include('scripts/pi-hole/lua/settings
                         <div class="form-group">
                             <div class="input-group">
                                 <div class="input-group-addon">Start</div>
-                                <input type="text" class="form-control DHCPgroup" id="dhcp.start" data-key="dhcp.start"
+                                <input type="text" class="form-control DHCPgroup"
                                     autocomplete="off" spellcheck="false" autocapitalize="none"
-                                    autocorrect="off" value="">
+                                    autocorrect="off" value="" placeholder="disabled" disabled>
                             </div>
                         </div>
                     </div>
@@ -45,9 +58,9 @@ mg.include('scripts/pi-hole/lua/settings
                         <div class="form-group">
                             <div class="input-group">
                                 <div class="input-group-addon">End</div>
-                                <input type="text" class="form-control DHCPgroup" id="dhcp.end" data-key="dhcp.end"
+                                <input type="text" class="form-control DHCPgroup"
                                     autocomplete="off" spellcheck="false" autocapitalize="none"
-                                    autocorrect="off" value="">
+                                    autocorrect="off" value="" placeholder="disabled" disabled>
                             </div>
                         </div>
                     </div>
@@ -56,9 +69,9 @@ mg.include('scripts/pi-hole/lua/settings
                         <div class="form-group">
                             <div class="input-group">
                                 <div class="input-group-addon">Router</div>
-                                <input type="text" class="form-control DHCPgroup" id="dhcp.router" data-key="dhcp.router"
+                                <input type="text" class="form-control DHCPgroup"
                                     autocomplete="off" spellcheck="false" autocapitalize="none"
-                                    autocorrect="off" value="">
+                                    autocorrect="off" value="" placeholder="disabled" disabled>
                             </div>
                         </div>
                     </div>
@@ -67,9 +80,9 @@ mg.include('scripts/pi-hole/lua/settings
                         <div class="form-group">
                             <div class="input-group">
                                 <div class="input-group-addon">Netmask</div>
-                                <input type="text" class="form-control DHCPgroup" id="dhcp.netmask" data-key="dhcp.netmask"
+                                <input type="text" class="form-control DHCPgroup"
                                     autocomplete="off" spellcheck="false" autocapitalize="none"
-                                    autocorrect="off" value="" placeholder="automatic">
+                                    autocorrect="off" value="" placeholder="disabled" disabled>
                             </div>
                         </div>
                     </div>
@@ -78,7 +91,7 @@ mg.include('scripts/pi-hole/lua/settings
                     </div>
                     <div class="col-md-12">
                         <div>
-                            <input type="checkbox" id="dhcp.ipv6" data-key="dhcp.ipv6" class="DHCPgroup">&nbsp;<label for="dhcp.ipv6"><strong>Enable additional IPv6 support (SLAAC + RA)</strong></label>
+                            <input type="checkbox" id="dhcp.ipv6" data-key="dhcp.ipv6" class="DHCPgroup" disabled>&nbsp;<label for="dhcp.ipv6"><strong>Enable additional IPv6 support (SLAAC + RA)</strong></label>
                             <p class="help-block">Enable this option to enable IPv6 support for the Pi-hole DHCP server. This will allow the Pi-hole to hand out IPv6 addresses to clients and also provide IPv6 router advertisements (RA) to clients. This option is only useful if the Pi-hole is configured with an IPv6 address.</p>
                         </div>
                     </div>
@@ -100,7 +113,7 @@ mg.include('scripts/pi-hole/lua/settings
                                 <div class="input-group-addon">Lease time</div>
                                 <input type="text" class="form-control DHCPgroup"
                                     autocomplete="off" spellcheck="false" autocapitalize="none"
-                                    autocorrect="off" id="dhcp.leaseTime" data-key="dhcp.leaseTime" value="">
+                                    autocorrect="off" value="disabled" disabled>
                             </div>
                         </div>
                         <p>The lease time can be in seconds, minutes (e.g., "45m"), hours (e.g., "1h"), days (like "2d"), or even weeks ("1w"). If no lease time is specified (empty), <code>dnsmasq</code>'s default lease time is one hour for IPv4 and one day for IPv6. You may also use "infinite" as string but be aware of the drawbacks: assigned addresses are will only be made available again after the lease time has passed or when leases are manually deleted below.</p>
@@ -109,13 +122,13 @@ mg.include('scripts/pi-hole/lua/settings
                 <div class="row">
                     <div class="col-sm-12">
                         <div>
-                            <input type="checkbox" id="dhcp.rapidCommit" data-key="dhcp.rapidCommit" class="DHCPgroup">&nbsp;<label for="dhcp.rapidCommit"><strong>Enable DHCPv4 rapid commit (fast address assignment)</strong></label>
+                            <input type="checkbox" id="dhcp.rapidCommit" data-key="dhcp.rapidCommit" class="DHCPgroup" disabled>&nbsp;<label for="dhcp.rapidCommit"><strong>Enable DHCPv4 rapid commit (fast address assignment)</strong></label>
                             <p class="help-block">The DHCPv4 rapid commit option allows the Pi-hole DHCP server to assign an IP address to a client right away. This can notably speed up the address assignment process and you will notice, e.g., faster WiFi joins in your network. This option should only be enabled if the Pi-hole DHCP server is the only DHCP server in your network.</p>
                         </div>
                     </div>
                     <div class="col-sm-12">
                         <div>
-                            <input type="checkbox" id="dhcp.multiDNS" data-key="dhcp.multiDNS" class="DHCPgroup">&nbsp;<label for="dhcp.multiDNS"><strong>Advertise DNS server multiple times</strong></label>
+                            <input type="checkbox" id="dhcp.multiDNS" data-key="dhcp.multiDNS" class="DHCPgroup" disabled>&nbsp;<label for="dhcp.multiDNS"><strong>Advertise DNS server multiple times</strong></label>
                             <p class="help-block">Advertise DNS server multiple times to clients. Some devices will add their own proprietary DNS servers to the list of DNS servers, which can cause issues with Pi-hole. This option will advertise the Pi-hole DNS server multiple times to clients, which should prevent this from happening.</p>
                         </div>
                     </div>
@@ -212,7 +225,7 @@ mg.include('scripts/pi-hole/lua/settings
         </div>
     </div>
     <div class="col-lg-12 settings-level-basic save-button-container">
-        <button type="button" class="btn btn-primary save-button"><i class="fa-solid fa-fw fa-floppy-disk"></i>&nbsp;Save & Apply</button>
+        <button type="button" class="btn btn-primary save-button" disabled><i class="fa-solid fa-fw fa-floppy-disk"></i>&nbsp;Save & Apply</button>
     </div>
 </div>
 <script src="<?=pihole.fileversion('scripts/vendor/bootstrap-toggle.min.js')?>"></script>
